### PR TITLE
Raise subclass of AttributeError in __getattr__

### LINKED ISSUE #986: hasattr not supported by _ParameterAttributeHandlers
**Describe the bug**

`hasattr` on a non-existent attribute of a `_ParameterAttributeHandler` subclass (e.g. LibraryChargeType) breaks in an `IndexError` instead of returning True/False.

**To Reproduce**
Steps to reproduce the behavior. A minimal reproducing set of python commands is ideal.

If the problem involves a specific molecule or file, please upload that as well.

```python
from openff.toolkit.typing.engines.smirnoff import ForceField
ff = ForceField('openff_unconstrained-1.3.0.offxml')
lc = ff.get_parameter_handler("LibraryCharges")
hasattr(lc.parameters[0], "charge2")
```

**Output**
The full error message (may be large, that's fine. Better to paste too much than too little.)

```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
openff/toolkit/typing/engines/smirnoff/parameters.py in __getattr__(self, item)
  1170             try:
-> 1171                 return indexed_attr_value[index]
  1172             except IndexError:

openff/toolkit/utils/collections.py in __getitem__(self, item)
   122             return self.__class__(super().__getitem__(item))
--> 123         return super().__getitem__(item)
   124 

IndexError: list index out of range

During handling of the above exception, another exception occurred:

IndexError                                Traceback (most recent call last)
<ipython-input-134-072f04b5c4f0> in <module>
----> 1 hasattr(lc.parameters[0], "charge2")

openff/toolkit/typing/engines/smirnoff/parameters.py in __getattr__(self, item)
  1171                 return indexed_attr_value[index]
  1172             except IndexError:
-> 1173                 raise IndexError(
  1174                     f"'{item}' is out of bound for indexed attribute '{attr_name}'"
  1175                 )

IndexError: 'charge2' is out of bound for indexed attribute 'charge'

```

**Computing environment (please complete the following information):**
- Operating system
- Output of running `conda list` 

**Additional context**

`hasattr` does not have its own dunder method, but rather relies on `getattr` raising an `AttributeError`. The relevant `getattr` method raises, variously, IndexErrors and KeyErrors as well as the AttributeError, so `hasattr` unexpectedly breaks. IMO any error raised in an `__getattr__` should be some version (e.g. a subclass) of `AttributeError` to preserve the functionality of this inbuilt function.

There may be other classes with this issue too, I haven't looked around.

https://github.com/openforcefield/openff-toolkit/blob/b2b76bb457c2a8235f86e887d75ee815030b605e/openff/toolkit/typing/engines/smirnoff/parameters.py#L1152-L1203

### PR DESCRIPTION
**Pull request by lilyminium**
Fixes #986 

- [x] Tag issue being addressed
- [x] Add [tests](https://github.com/openforcefield/openff-toolkit/tree/master/openff/toolkit/tests)
- [ ] Update docstrings/[documentation](https://github.com/openforcefield/openff-toolkit/tree/master/docs), if applicable
- [ ] [Lint](https://open-forcefield-toolkit.readthedocs.io/en/latest/developing.html#style-guide) codebase
- [x] Update [changelog](https://github.com/openforcefield/openff-toolkit/blob/master/docs/releasehistory.rst)

### CODE REVIEW

File: openff/toolkit/utils/exceptions.py [hunk 1]
```diff
@@ -290,3 +290,9 @@ class DuplicateVirtualSiteTypeException(OpenFFToolkitException):
 
 class CallbackRegistrationError(OpenFFToolkitException, TypeError):
     """Error raised when callback registration fails."""
+
+
+class MissingIndexedAttributeError(
+    OpenFFToolkitException, IndexError, KeyError, AttributeError
+):
+    """Error raised when an indexed attribute does not exist"""
```
Comment (mattwthompson) [inline openff/toolkit/utils/exceptions.py:298]: Non-blocking: We should plan to remove the inheritance from `IndexError` and `KeyError` at some point but I haven't thought about how to communicate/implement this. I wouldn't want a message warning users to not catch as `IndexError`/`KeyError` if they're not doing that, or even close to that specific case. If it's not tractable it's probably okay to just leave the other errors in long-term.

**Review by mattwthompson — APPROVED** [review]
I was pretty confused my some of the behavior of these attributes but then I realized they used the descriptor pattern, which has tripped me up in the past. Looks like it does what you wanted in the bug report:

```
>>> from openff.toolkit.typing.engines.smirnoff import ForceField
/Users/mwt/software/openforcefield/openff/toolkit/utils/exceptions.py:18: DeprecationWarning: ParseError is DEPRECATED and will be removed in a future release of the OpenFF Toolkit.
  warnings.warn(warning_msg, DeprecationWarning)
>>> ff = ForceField('openff_unconstrained-1.3.0.offxml')
>>> lc = ff.get_parameter_handler("LibraryCharges")
>>> hasattr(lc.parameters[0], "charge1")
True
>>> hasattr(lc.parameters[0], "charge2")
False
>>> hasattr(lc.parameters[0], "charge100")
False
>>> getattr(lc.parameters[0], "charge2")
Traceback (most recent call last):
  File "/Users/mwt/software/openforcefield/openff/toolkit/typing/engines/smirnoff/parameters.py", line 1105, in __getattr__
    return indexed_attr_value[index]
  File "/Users/mwt/software/openforcefield/openff/toolkit/utils/collections.py", line 119, in __getitem__
    return super().__getitem__(item)
IndexError: list index out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mwt/software/openforcefield/openff/toolkit/typing/engines/smirnoff/parameters.py", line 1107, in __getattr__
    raise MissingIndexedAttributeError(
openff.toolkit.utils.exceptions.MissingIndexedAttributeError: 'charge2' is out of bounds for indexed attribute 'charge'
>>> getattr(lc.parameters[0], "charge1")
Quantity(value=1.0, unit=elementary charge)
>>> getattr(lc.parameters[0], "charge100")
Traceback (most recent call last):
  File "/Users/mwt/software/openforcefield/openff/toolkit/typing/engines/smirnoff/parameters.py", line 1105, in __getattr__
    return indexed_attr_value[index]
  File "/Users/mwt/software/openforcefield/openff/toolkit/utils/collections.py", line 119, in __getitem__
    return super().__getitem__(item)
IndexError: list index out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mwt/software/openforcefield/openff/toolkit/typing/engines/smirnoff/parameters.py", line 1107, in __getattr__
    raise MissingIndexedAttributeError(
openff.toolkit.utils.exceptions.MissingIndexedAttributeError: 'charge100' is out of bounds for indexed attribute 'charge'
```

**Comment by j-wags** [comment]
Any blockers to merging? This should be good to go once releasenotes are updated.

**Comment by lilyminium** [comment]
> Any blockers to merging? 

@j-wags not that I know of

**Pull request merged**